### PR TITLE
Tree-sitter grammar: support DB type reference

### PR DESCRIPTION
**Pull request by OceanOak**
Changelog:

```
Tree-sitter-darklang
- Update the grammar to support DB type reference
```

#5321

### CODE REVIEW

File: packages/darklang/languageTools/parser/identifiers.dark [hunk 1]
```diff
@@ -54,14 +54,43 @@ module Darklang =
           (node: ParsedNode)
           : Stdlib.Result.Result<WrittenTypes.QualifiedTypeIdentifier, String> =
           if node.typ == "qualified_type_name" then
+            let filteredChildren =
+              node.children
+              |> Stdlib.List.filter (fun n ->
+                n.typ != "type_args"
+                && n.fieldName != Stdlib.Option.Option.Some "symbol_open_angle"
```
Comment (OceanOak) [inline packages/darklang/languageTools/parser/identifiers.dark:61]: not very happy with this solution, any suggestion of alternatives please?

File: packages/darklang/prettyPrinter/programTypes.dark [hunk 2]
```diff
@@ -883,7 +883,7 @@ module Darklang =
           | [] -> ""
           | _ ->
             p.declaration.typeParams
-            |> Stdlib.List.map (fun typeParam -> $"'{typeParam}")
+            |> Stdlib.List.map (fun typeParam -> $"{typeParam}")
```
Comment (StachuDotNet) [inline packages/darklang/prettyPrinter/programTypes.dark:886]: Hm - let's update WrittenTypes to record the `'` as a symbol, separate from the actual type variable.
Then, the pretty-printer probably needs the `'` to be included like the source was.

File: tree-sitter-darklang/grammar.js [hunk 3]
```diff
@@ -70,9 +70,26 @@ module.exports = grammar({
       seq(
         field("keyword_type", alias("type", $.keyword)),
         field("name", $.type_identifier),
+        optional(
+          seq(
+            field("symbol_open_angle", alias("<", $.symbol)),
+            field("type_params", $.type_decl_type_params),
+            field("symbol_close_angle", alias(">", $.symbol)),
+          ),
+        ),
```
Comment (StachuDotNet) [inline tree-sitter-darklang/grammar.js:79]: the syntax to apply type args will be the same for many places (`DB<Int>`, `myFn<Int>`, `MyType<Int>`) - let's extract this out to a `type_args` root-level thing, and share across the use cases.

File: tree-sitter-darklang/grammar.js [hunk 4]
```diff
@@ -874,8 +903,25 @@ module.exports = grammar({
...
+      ),
+
+    type_args: $ =>
+      seq(
+        $.type_reference,
+        repeat(
+          seq(field("symbol_comma", alias(",", $.symbol)), $.type_reference),
+        ),
       ),
```
Comment (StachuDotNet) [inline tree-sitter-darklang/grammar.js:925]: Separate from my other long comment: I do think that `type_args` should include the `<` symbols.

File: packages/darklang/languageTools/parser/identifiers.dark (hunk 1 above)
Comment (StachuDotNet) [reply packages/darklang/languageTools/parser/identifiers.dark:61]: I think `type_args` should include the `<` and `>` characters - then, the task seems simpler: make sure that `<` is the first thing in the list, `>` is the last thing in the list, and then extract the other things as you're doing. Then, you just need to look for the one `type_args` node, here.

File: tree-sitter-darklang/test/corpus/exhaustive/type_refs.txt
Comment (StachuDotNet) [inline tree-sitter-darklang/test/corpus/exhaustive/type_refs.txt:1]: the titles of the first 2 tests in this file are reversed (1st title corresponds to 2nd test, and vice versa)

File: tree-sitter-darklang/grammar.js [hunk 5]
```diff
@@ -70,10 +71,28 @@ module.exports = grammar({
...
         field("typ", $.type_decl_def),
       ),
 
+    type_params: $ =>
+      seq(
+        field("symbol_open_angle", alias(token.immediate("<"), $.symbol)),
+        field("params", $.params),
+        field("symbol_close_angle", alias(token.immediate(">"), $.symbol)),
+      ),
+    params: $ =>
```
Comment (StachuDotNet) [inline tree-sitter-darklang/grammar.js:85]: This name is a bit broad - maybe rename to something like `type_params_items` or inline it in the `type_params` rule?

File: tree-sitter-darklang/test/corpus/exhaustive/type_refs.txt
Comment (OceanOak) [reply tree-sitter-darklang/test/corpus/exhaustive/type_refs.txt:1]: `all builtin type references` and` (simple) qualified type identifier` ?
maybe wrong file?

Comment (StachuDotNet) [reply tree-sitter-darklang/test/corpus/exhaustive/type_refs.txt:1]: Sorry - I meant the lines 263 and 330 in this file should be swapped, I believe

Comment (OceanOak) [reply tree-sitter-darklang/test/corpus/exhaustive/type_refs.txt:1]: yes, thanks!

**Pull request merged**